### ISSUE TITLE
Generate Enums in separate classes

### ISSUE THREAD

**Issue by sauravsarangi** [joelittlejohn/jsonschema2pojo#1002]
Hi,

Currently the enums are getting generated as an inner-class definition inside the referencing object. Has there been any feature added later which creates the class files for the enums separately and only are referenced through a property in the referencing object?

**Comment by ddcruver** [comment]
Try using javaType property. Ex.) "javaType": "com.example.MyEnumType".

**Comment by jacobhampel** [comment]
I tried to add the javaType to the enum but instead of generating a Java enum a java class was generated.
The JSON schema: 
```
"HashAlgorithmEnumType": {
      "javaType": "HashAlgorithmEnum",
      "type": "string",
      "additionalProperties": false,
      "enum": [
        "SHA256",
        "SHA384",
        "SHA512"
      ]
    }
```
The resuling java class begins with
```
public class HashAlgorithmEnum {
```
instead of 
```
public enum HashAlgorithmEnum {
```

**Comment by bluecapz** [comment]
@jacobhampel , since you're defining enum, try removing the **type** & **additionalProperties** attribute as I think that would make it an object already. Just leave only **enum** and **javaType**.

```
"HashAlgorithmEnumType": {
      "javaType": "HashAlgorithmEnum",
      "enum": [
        "SHA256",
        "SHA384",
        "SHA512"
      ]
    }
```

**Comment by ddcruver** [comment]
I would expect that to work. Can you try to break the enumeration into a separate file and then use $ref to reference it?

**Comment by bluecapz** [comment]
Yep. Currently using `$ref` for a string field restricting value as defined in enum. Something like this:

```
"fruits": {
    "description": "Field in an object expecting an enum value",
    "$ref": "#/definitions/fruit",
}

...

"definitions": {
    "fruit": {
        "description": "Fruit enum in separate class",
        "enum": ["Apple", "Banana", "Cherry"],
        "javaType": "my.package.Fruit"
    }

}
```

**Comment by jacobhampel** [comment]
@bluecapz I tried your suggestion but the generator generated a "class" instead of an "enum". 
Here is the full JSON-schema file:
```
{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "id": "urn:draft:GenericResponse",
    "title": "GenericResponse",
    "type": "object",
    "properties": {
        "status": {
            "javaType": "GenericStatusResponse",
            "enum": [
                "Accepted",
                "Rejected"
            ]
        }
    },
    "additionalProperties": false,
    "required": [
        "status"
    ]
}
```
I tried the online version of jsonschema2pojo to make sure that I'm using the latest version.
If I remove the "javaType" property from the enum in the JSON-schema a nested but correct enum is generated. 
This has the main disadvantage that I can't simply reuse the enum across multiple java classes.

**Comment by ddcruver** [comment]
@jacobhampel There might be an actual bug.  If you provide a non-default java package it works correctly.

```json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "id": "urn:draft:GenericResponse",
  "title": "GenericResponse",
  "javaType": "org.example.GenericResponse",
  "type": "object",
  "properties": {
    "status": {
      "javaType": "org.example.GenericStatusResponse",
      "enum": [
        "Accepted",
        "Rejected"
      ]
    }
  },
  "additionalProperties": false,
  "required": [
    "status"
  ]
}
```

**Comment by bluecapz** [comment]
@jacobhampel , not so sure but probably the reason why it is generated as a class is because it is embedded in the property definition itself, where the containing definition is defined as an `object`. I'm not entirely sure but what I did to have a separate enum that can be reused also (via `$ref` references) is to put the enum into a separate definition and I can just reuse it anytime. 
The disadvantage of embedding it directly into a property is that it can't be easily referenced by other definitions (or at least how I think about it). While having it in a separate definition, would mean a separate & independent enum that can be reused/referenced by other classes as well. 

Also not entirely sure about default packaging as mentioned by @ddcruver but I would be rewriting it to like this:

```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "id": "urn:draft:GenericResponse",
  "title": "GenericResponse",
  "javaType": "org.example.GenericResponse",
  "type": "object",
  "properties": {
    "status": {
      "$ref": "#/definitions/status"
  },
  "additionalProperties": false,
  "required": [
    "status"
  ],
  "definitions": {
    "status" : {
        "enum": ["Accepted", "Rejected"],
        "javaEnumNames": ["STATUS_ACCEPTED", "STATUS_REJECTED"],
        "javaType": "org.example.GenericStatusResponse"
    }
  }
}
```

But if you intend to have that enum as a single schema definition, you might want to try having it like this:
```
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "id": "urn:draft:GenericResponse",
  "title": "GenericResponse",
  "javaType": "org.example.GenericResponse",
  "enum": ["Accepted", "Rejected"],
  "javaEnumNames": ["STATUS_ACCEPTED", "STATUS_REJECTED"],
  "javaType": "org.example.GenericStatusResponse"
}
```
I haven't tried this since I haven't tried multiple schema files. But I assume this would work and would generate an `Enum` rather than a class since `"type": "object"` is not present as I've previously pointed out also.

**Comment by jacobhampel** [comment]
Indeed. It works like a charm.
One small hint: You have to add an `}` right bevor the `, additionalProperties`

**Comment by sergeykad** [comment]
It is not always feasible to extract Enum to a separate JSON Schema file. For example if a client defined Schema and you cannot change it. 
An option in the plugin to generate all classes in separate files can be helpful.

**Comment by marceloverdijk** [comment]
Is it possible to generate an enum in a separate class without adding the `javaType`?

I'm having a `Gender` enum like:

```
    "Gender": {
      "type": "string",
      "enum": ["MALE", "FEMALE"]
    }
```

which is referenced in many places.

(I cannot change the json schema)

**Comment by winterkid** [comment]
Hi, this feature would be quite useful to us. Especially in use cases where we are bound to use a third party schema and thus cannot enhance or change it - e.g. with javaType. Is there any progress on this issue?

**Comment by adrien1212** [comment]
As the solution is not implemented I provide a temporary solution :

1. Create a CustomRuleFactory
2. Override the method getSchemaRule
3. CustomSchemaRule must extends SchemaRule and override apply()

### 1. and 2.
```java
public class CustomRuleFactory  extends RuleFactory {
    @Override
    public Rule<JClassContainer, JType> getSchemaRule() {
    	return new CustomSchemaRule(this);
    }
}
```

### 3.
In the method `apply()` replace the following line
```java
public class CustomSchemaRule extends SchemaRule {
   @Override
    public JType apply(String nodeName, JsonNode schemaNode, JsonNode parent, JClassContainer generatableType, Schema schema) {

    if (schemaNode.has("enum")) {
            javaType = ruleFactory.getEnumRule().apply(nodeName, schemaNode, parent, generatableType, schema);

    // BY 

    if (schemaNode.has("enum")) {
            javaType = ruleFactory.getEnumRule().apply(nodeName, schemaNode, parent, generatableType.getPackage(), schema);
}
```

The enumeration is now genereted in a separate class

**Comment by Cs4r** [comment]
> [@jacobhampel](https://github.com/jacobhampel) , since you're defining enum, try removing the **type** & **additionalProperties** attribute as I think that would make it an object already. Just leave only **enum** and **javaType**.
> 
> ```
> "HashAlgorithmEnumType": {
>       "javaType": "HashAlgorithmEnum",
>       "enum": [
>         "SHA256",
>         "SHA384",
>         "SHA512"
>       ]
>     }
> ```

this worked for me
